components/target_estimator.py

```diff
@@ -0,0 +1,102 @@
+from components.chassis import Chassis
+from components.range_finder import RangeFinder
+from components.turret import Turret
+from components.vision import Vision, VisionData
+
+from utilities.functions import constrain_angle
+
+import math
+import time
+
+from typing import Optional
+
+
+class TargetEstimator:
+
+    ALPHA = 0.1  # Exponential smoothing fraction
+    CAMERA_TO_LIDAR = 0.15
+
+    chassis: Chassis
+    range_finder: RangeFinder
+    turret: Turret
+    vision: Vision
+
+    def __init__(self) -> None:
+        self.reset()
+
+    def on_enable(self) -> None:
+        self.reset()
+
+    def reset(self) -> None:
+        self.angle_to_target: float = 0.0
+        self.vision_range: float = 0.0
+        self.previous_azimuth: float = 0.0
+        self.previous_heading: Optional[float] = None
+        self.previous_vision_data: VisionData = VisionData(0.0, 0.0, 0.0)
+        self._pointing_downrange: bool = False
+
+    def is_ready(self) -> bool:
+        return self.is_pointing_downrange() and self.has_vision()
+
+    def has_vision(self) -> bool:
+        return self.vision.is_ready()
+
+    def is_pointing_downrange(self) -> bool:
+        return self._pointing_downrange
+
+    def get_data(self) -> VisionData:
+        if abs(self.angle_to_target) < math.radians(5.0):
+            vd = VisionData(
+                self.range_finder.get_distance() - self.CAMERA_TO_LIDAR,
+                self.angle_to_target,
+                time.monotonic(),
+            )
+        else:
+            vd = VisionData(self.vision_range, self.angle_to_target, time.monotonic(),)
```

Comment:
If we can't see a target, this returns bad data. See my PR comment for an alternative.

Reply:
I'm just worried about turning this into a full blown state estimator. Once we start fusing odometry it gets really complicated and I think we'd be better off using a proper ukf to do it. That means tuning a ukf...
I do want to be able to maintain the estimate when we lose sight during an unwind of the turret - this currently just resets the filter and starts again. I can't see a nice way to do that until we have the turret logic moved out into your turret controller.

Reply:
I'm just worried about turning this into a full blown state estimator. Once we start fusing odometry it gets really complicated and I think we'd be better off using a proper ukf to do it. That means tuning a ukf...
I do want to be able to maintain the estimate when we lose sight during an unwind of the turret - this currently just resets the filter and starts again. I can't see a nice way to do that until we have the turret logic moved out into your turret controller.

Reply:
I don't follow. If we know our position and heading on the field and the azimuth of our turret, we can estimate the azimuth of the target because we know where it is on the field. It isn't fusing odometry any more than you are already doing. It doesn't need to be particularly accurate, just enough to get us back to where we can see the target. The current plan is to merge my turret changes, then pull this PR and make these tweaks.

Reply:
We don't know our position on the field very well. We initialise the chassis with something close to the truth, given the accuracy of the drive team placing the robot. We then dead reckon our way around the field and the error accumulates, and will get especially bad when we drive over the metal in the middle of the field. I think using odometry without any way to correct our absolute position will cause problems.

Reply:
Ok. I've just changed the scan start point to be just plain downrange, i.e. the negative of the heading. That's very coarse but no worse than the corrections the target estimator already does, and it should be good enough to start scanning in the right general vicinity.